### PR TITLE
Changes to MessageEvent

### PR DESCRIPTION
**Pull request by windy1**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1111) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/548) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/229) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/559)

This PR makes some substantial changes to how MessageEvents are handled. Instead of passing around a `Text` object, the event now contains a `MessageFormatter` which can be used to modify the event's message in a more reliable and modular fashion. This will make interactions and compatibility between chat plugins a lot easier. Previous methods like `getOriginalMessage`, `getRawMessage`, `getMessage`, and `setMessage` still behave as expected except instead of Optionals, a non-null Text is always returned. Developers should use `setMessageCancelled` to cancel the sending instead of setting the message to null.

----

Some example uses:

Let's say I would like to insert a "group" prefix to the message but I also would like to keep existing prefixes that other plugins may or may not have already applied. I can do something like:

```java
class GroupApplier extends SimpleTextTemplateApplier {
    public GroupApplier() {
        super(of('[', arg("group"), "] "));
    }
}
```

```java
@Listener
void onChat(MessageChannelEvent.Chat event) {
    GroupApplier applier = new GroupApplier();
    applier.setParameter("group", Text.builder("admin").color(TextColors.RED));
    event.getFormatter().getHeader().insert(0, applier);
}
```

On an unaltered message, this will produce something equivalent to:

`[<red>admin</red>] <NAME> MESSAGE`

Now lets say another plugin wants to override that prefix and make it say "mod" instead of "admin". Well then they can do:

```java
event.getHeader().forEach(GroupApplier.class, applier -> applier.setParameter("group", Text.builder("mod").color(TextColors.AQUA));
```

These are only a few of the many uses this PR will provide.

**Note:** This PR depends on https://github.com/SpongePowered/SpongeAPI/pull/1079. https://github.com/SpongePowered/SpongeAPI/pull/1079 should be merged into master *and then into bleeding* before this PR is merged.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/entity/living/humanoid/player/KickPlayerEvent.java [hunk 1]
```diff
@@ -25,11 +25,14 @@
...
 import org.spongepowered.api.event.message.MessageChannelEvent;
+import org.spongepowered.api.eventgencore.annotation.ImplementedBy;
 
 /**
  * Fired when a {@link Player} is kicked.
  */
+@ImplementedBy(AbstractMessageEvent.class)
 public interface KickPlayerEvent extends TargetPlayerEvent, MessageChannelEvent {
 
 }
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/entity/living/humanoid/player/KickPlayerEvent.java]: Add a newline since you're editing this file anyway?

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 2]
```diff
@@ -24,37 +24,254 @@
...
+
+    /**
+     * Parameter for body section in body partition.
+     */
+    String PARAM_MESSAGE_BODY = "body";
+
+    /**
+     * Parameter for footer section in footer partition.
+     */
+    String PARAM_MESSAGE_FOOTER = "footer";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: "footer" doesn't make sense here, "suffix" does

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 3]
```diff
@@ -24,37 +24,254 @@
...
 /**
  * Describes events when a involving a {@link Text} message.
  */
+@ImplementedBy(AbstractMessageEvent.class)
 public interface MessageEvent extends Event {
 
     /**
+     * Parameter for header section in header partition.
+     */
+    String PARAM_MESSAGE_HEADER = "header";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: "header" doesn't make sense here, "prefix" does

File: src/main/java/org/spongepowered/api/service/permission/Subject.java [hunk 4]
```diff
@@ -113,12 +112,15 @@
...
      * Test whether the subject is permitted to perform an action given as the
-     * given permission string.
+     * given permission string. This must return the same value as
+     * {@link #hasPermission(Set, String)} using {@link #getActiveContexts()}.
      *
      * @param permission The permission string
      * @return True if permission is granted
      */
-    boolean hasPermission(String permission);
+    default boolean hasPermission(String permission) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/service/permission/Subject.java]: Are these changes intentional? They seem unrelated to this PR.

**Comment by kashike** [comment]
It's quite hard to tell what's part of this PR and what's from #1079, so I apologise if any comments are duplicates.

File: src/main/java/org/spongepowered/api/service/permission/Subject.java (hunk 4 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/service/permission/Subject.java]: These are @zml2008's changes. It's displayed in the diff because I branched off of bleeding, merged master into my new branch, and then applied my changes.

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java (hunk 2 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: I disagree, footer is more general than suffix. You can add "suffixes" to the "footer" but that doesn't make the whole footer a suffix.

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java (hunk 3 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: Same as above.

**Comment by windy1** [comment]
@kashike Look at the individual commits instead of the overall diff. This is a result of merging master into my bleeding branch.
